src/Microsoft.Diagnostics.Monitoring.WebApi/Stacks/EventStacksPipeline.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring.EventPipe;
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Tracing;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.WebApi.Stacks
+{
+    internal sealed class EventStacksPipelineSettings : EventSourcePipelineSettings
+    {
+        public EventStacksPipelineSettings()
+        {
+            Duration = System.Threading.Timeout.InfiniteTimeSpan;
+        }
+
+        public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(5);
+    }
+
+    internal sealed class EventStacksPipeline : EventSourcePipeline<EventStacksPipelineSettings>
+    {
+        private TaskCompletionSource<StackResult> _stackResult = new(TaskCreationOptions.RunContinuationsAsynchronously);
+        private StackResult _result = new();
+
+        public EventStacksPipeline(DiagnosticsClient client, EventStacksPipelineSettings settings)
+            : base(client, settings)
+        {
+        }
+
+        protected override MonitoringSourceConfiguration CreateConfiguration()
+        {
+            return new EventPipeProviderSourceConfiguration(requestRundown: false, bufferSizeInMB: 256, new[]
+            {
+                new EventPipeProvider(StackEvents.Provider, System.Diagnostics.Tracing.EventLevel.LogAlways)
+            });
+        }
+
+        protected override async Task OnEventSourceAvailable(EventPipeEventSource eventSource, Func<Task> stopSessionAsync, CancellationToken token)
+        {
+            eventSource.Dynamic.AddCallbackForProviderEvents((string provider, string _) => provider == StackEvents.Provider ?
```

Comment:
You can use `AddCallbackForProviderEvents(string providerName, string eventName, Action<TraceEvent> callback)` by passing `null` for the `eventName` parameter.